django/db/models/sql/where.py

```diff
@@ -35,48 +35,87 @@ class WhereNode(tree.Node):
     resolved = False
     conditional = True
 
-    def split_having(self, negated=False):
+    def must_remain_connected(self, negated=False):
         """
-        Return two possibly None nodes: one for those parts of self that
-        should be included in the WHERE clause and one for those parts of
-        self that must be included in the HAVING clause.
+        Whether or not children must be connected in the same filtering clause
+        (WHERE > HAVING > QUALIFY) to maintain logical semantic.
         """
-        if not self.contains_aggregate:
-            return self, None
         in_negated = negated ^ self.negated
-        # If the effective connector is OR or XOR and this node contains an
-        # aggregate, then we need to push the whole branch to HAVING clause.
-        may_need_split = (
+        return (
             (in_negated and self.connector == AND)
             or (not in_negated and self.connector == OR)
             or self.connector == XOR
         )
-        if may_need_split and self.contains_aggregate:
-            return None, self
+
+    def split_having_qualify(self, negated=False, must_group_by=False):
+        """
+        Return three possibly None nodes: one for those parts of self that
+        should be included in the WHERE clause, one for those parts of
+        self that must be included in the HAVING clause, and one for those
+        parts that refer to window functions.
+        """
+        if not self.contains_aggregate and not self.contains_over_clause:
+            return self, None, None
+        in_negated = negated ^ self.negated
+        must_remain_connected = self.must_remain_connected(negated)
```

Comment:
`must_remain_connected()` is called only here maybe it's better to pass `in_negated` instead of `negated` and don't calculate it twice :thinking:

Reply:
There's no need for a separate method anymore; I've merged it back in `split_having_qualify`.